### PR TITLE
fix(topic): apply explicit zero queue counts when updating a topic

### LINKED ISSUE #4582: [Bug] Topic update silently ignores explicit zero queue counts
### Before Creating the Bug Report

- [x] I have searched the [open issues](https://github.com/apache/rocketmq-dashboard/issues) of this repository and believe that this is not a duplicate.

- [x] This is a defect in RocketMQ Studio, not a usage question and not a defect in another Apache RocketMQ repository.

- [x] I can reproduce this on the current `master` branch, or I have stated the exact version I am running below.

### Studio Version

branch: master
git commit id: d50ffecc
deployed as: built from source

### Runtime Environment

OS: Ubuntu 24.04 (WSL2)
MySQL: 8.x, container from deploy/docker-compose.yml
browser: not applicable (API-level defect)

### Connected RocketMQ Cluster

Not applicable: reproduced with a unit test (`RocketMQAdminClientImplTest`) on master, where `DefaultMQAdminExt` is mocked. The defect is in the update path itself, before any cluster call.

### Build Toolchain

_No response_

### Describe the Bug

`POST /api/topics/update` (and the `rmq.topic.update` AI tool) silently ignores an explicit `writeQueues: 0`. A write-queue count of 0 is a legitimate way to stop producers while consumers drain a topic, but the request behaves exactly like omitting the field: the stored queue count (or the default of 8) is applied instead.

Root cause: `TopicVO` declares `writeQueues`/`readQueues` as primitive `int`, so once `UpdateTopicDTO.toTopicVO()` / `TopicUpdateInput.mergeWith()` copy the value, an explicit 0 and an omitted field are indistinguishable. `RocketMQAdminClientImpl.updateTopic` then uses a `> 0` check to mean "not provided".

### Steps to Reproduce

1. On master, run the (new) unit test `RocketMQAdminClientImplTest.updateTopicAppliesExplicitZeroQueueCounts`, or:
2. Call `POST /api/topics/update` for an existing 8/8 topic with body `{"instanceId": "...", "name": "orders", "writeQueues": 0, "readQueues": 8}`.
3. The broker config and the stored metadata keep `writeQueueNums: 8`; the response also reports 8.

### What Did You Expect to See?

An explicit `writeQueues: 0` is applied to the broker topic config and the stored metadata. An omitted value continues to preserve the stored count (the fix for #4123/#4275 already established "absent keeps the stored value").

### What Did You See Instead?

`writeQueues: 0` is treated as "not provided" and silently replaced with the previous value or the default of 8.

### Additional Context

Suggested direction: make the two `TopicVO` fields nullable (`Integer`), mirroring the optional semantics `perm` already uses, and let the Apache update path apply an explicit 0. I have a failing-test-first patch ready and will link the PR here shortly.

### Are You Willing to Submit a Pull Request?

- [x] Yes, I am willing to submit a pull request.

### PR DESCRIPTION
**Pull request by Frun1na**
fix(topic): apply explicit zero queue counts when updating a topic

### Which Issue(s) This PR Fixes

- Fixes #4582

### Brief Description

An update that submitted `writeQueues: 0` — a legitimate way to stop producers
while consumers drain a topic — was silently ignored: the stored or default
queue count was applied instead. `TopicVO` held `writeQueues`/`readQueues` as
primitive ints, so the provider could not tell an explicit 0 from a value the
caller never sent, and its `> 0` check treated both as "not provided".

This makes the two `TopicVO` fields nullable (`Integer`), mirroring the
optional semantics `perm` already uses:

- the Apache update path applies an explicit 0 unchanged, while an omitted
  value still preserves the stored count (existing behaviour, covered by
  `updateTopicPreservesQueueCountsWhenNotSpecified`);
- the create path is unchanged (0 still falls back to the default of 8);
- Tencent validation keeps its current semantics (null behaves like the old 0);
- the web topic form keeps `min={1}`, so this change is observable over the
  REST API and the AI tools today; relaxing the form is left for a separate
  discussion.

### How Did You Test This Change?

`cd server && mvn -B -ntp test` — `Tests run: 2415, Failures: 0, Errors: 0, Skipped: 0`, BUILD SUCCESS.

- New regression test `updateTopicAppliesExplicitZeroQueueCounts` fails on
  master (`expected: 0 but was: 8`) and passes with this change.
- The existing queue-count tests (`updateTopicPreservesQueueCountsWhenNotSpecified`,
  `updateTopicAppliesExplicitQueueCounts`) still pass, so partial updates and
  non-zero resizes behave exactly as before.

### Checklist

- [x] One coherent change; unrelated modifications are not bundled in
- [x] Commit subject follows Conventional Commits (`feat:` / `fix:` / `refactor:` / `chore:` / `docs:` / `perf:`)
- [x] Tests added or updated for non-trivial changes, test methods named `...Test`
- [ ] New UI text has both Chinese and English entries under `web/src/i18n/`
- [x] Architecture constraints stay green (`mvn test` runs the ArchUnit checks)
- [x] New source files carry the ASF license header
- [ ] Documentation touched where behaviour changed (README / `docs/` / in-app help)

### CODE REVIEW

File: server/src/main/java/org/apache/rocketmq/studio/provider/apache/RocketMQAdminClientImpl.java [hunk 1]
```diff
@@ -359,11 +361,11 @@ public TopicVO updateTopic(String instanceId, TopicVO topic) {
                 // Preserve the existing queue counts when the update request does not change them,
                 // matching the perm semantics below; defaulting to 8 would silently resize the
                 // topic on partial updates (e.g. perm or remark only).
-                int writeQueues = topic.getWriteQueues() > 0
+                int writeQueues = topic.getWriteQueues() != null && topic.getWriteQueues() >= 0
                         ? topic.getWriteQueues()
                         : existing != null && existing.getWriteQueueNums() != null
                                 && existing.getWriteQueueNums() > 0 ? existing.getWriteQueueNums() : 8;
```
[the remark this answers is not in the document]
Comment (Frun1na) [reply server/src/main/java/org/apache/rocketmq/studio/provider/apache/RocketMQAdminClientImpl.java]: Good catch — the fallback would indeed have undone a stored drained state on the next partial update. Both occurrences now accept a stored zero (`>= 0`), covered by a regression test (`updateTopicPreservesStoredZeroQueueCountsWhenOmitted`) asserting that an update omitting the counts keeps stored write=0/read=8.

**Review by RockteMQ-AI — APPROVED** [review]
﻿### Summary
This PR fixes a semantic issue where `writeQueues` and `readQueues` were primitive `int` in `TopicVO`, making it impossible to distinguish "not submitted" (should preserve existing value) from "explicitly set to 0" (a legitimate drain operation). The fix changes them to `Integer` (nullable) and updates all null-check logic in both Apache and Tencent providers.

The change is well-designed:
- `null` → not submitted → preserve existing value
- `0` → explicit drain → apply zero
- `> 0` → set specific count

The `>= 0` check correctly addresses the edge case where a stored zero should survive partial updates (Copilot's feedback has been properly addressed). The regression tests cover both the explicit-zero update and the stored-zero preservation scenarios.

One minor note: the `createTopic` path still defaults to 8 when `writeQueues`/`readQueues` is null or ≤ 0, which is correct for creation semantics.

LGTM.

---
*Automated review by github-manager-bot*

**Comment by lizhimins** [comment]
Thanks — the underlying problem you identified is real, and the two tests you added are both mutation-sensitive.

To restate it so we are clear about what is being decided: `UpdateTopicDTO` already carries `writeQueues`/`readQueues` as `Integer` with `@PositiveOrZero` and only copies them onto the VO when non-null, but `TopicVO` declares them as primitive `int`, so "not submitted" arrives at `RocketMQAdminClientImpl.updateTopic` as `0` and the `> 0` test reads it as "not provided". An explicit drain-to-zero is therefore dropped.

**We are not taking this approach.** Our position is that the update path should use the value the caller passed, as passed. This PR goes the other way: it makes the VO nullable and then resolves the count through a three-way substitution at every consumer — request value, else the stored `RmqTopic` value, else a hardcoded 8. Each of those resolution points is a place where the wrong number can be written to the broker, and the PR widens the number of consumers that have to get the resolution right rather than removing the ambiguity at the boundary. Concretely:

1. **The nullable VO leaks into contracts that were not updated.** Three AI output records still take a primitive `int` and are constructed straight from the VO: `TopicOutput`, `TopicListItem` and `TopicDetailOutput`. We traced every caller and none can pass null today — but this PR is what makes null possible, and the guards are caller discipline rather than construction. That is exactly the "someone changes the wrong one later" failure mode.
2. **It changes the public REST response shape, which the description does not say.** `AliyunInstanceProvider.createTopic` and `updateTopic` return the caller's `TopicVO` verbatim and never populate the queue counts, and `TencentInstanceProvider.updateTopic` only writes them back when a queue number was requested. So a remark-only `POST /api/topics/update` against a cloud instance would start serializing `"writeQueues": null` where trunk returns `0`. `web/src/api/metadata.ts` declares `writeQueues: number` as non-nullable and `topic.tsx` renders it directly. Note also that the topic form hides the queue inputs entirely for cloud instances, so the UI never sends them there.
3. **The second commit carries an undisclosed behaviour change.** Flipping `existing.getWriteQueueNums() > 0` to `>= 0` means a *stored* 0 now survives a partial update instead of being reset to 8. That is an independent fix from the request-side one, but the description presents it as "existing behaviour".

The drain-to-zero case is also only reachable from non-UI callers: the topic edit form declares the counts `required` with `min={1}`, so Studio itself cannot submit 0.

**What would make a future attempt welcome.** Keep `TopicVO` non-nullable and settle the ambiguity at the API boundary instead, so the value that reaches the broker is the value the caller sent: make the counts required on the update DTO (they are already `Integer` + `@PositiveOrZero`) and confine the default-8 substitution to `createTopic`, where it belongs. If you would rather keep partial updates, then the omitted case needs a representation that does not depend on every downstream consumer remembering to null-check.

One mechanical note either way: #4538 has landed, so `RocketMQAdminClientImpl.updateTopic` now throws 404 for a missing topic just above the queue-count block. This branch needs a rebase onto `rocketmq-studio`, and the `existing != null &&` guards become redundant once you do.
